crates/transcode/src/env_check.rs

```diff
@@ -0,0 +1,422 @@
+use ink_metadata::InkProject;
+use scale_info::{
+    form::PortableForm,
+    Field,
+    PortableRegistry,
+    TypeDef,
+};
+
+use anyhow::{
+    Context,
+    Result,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum EnvCheckError {
+    CheckFailed(String),
+    RegistryError(String),
+}
+
+impl std::fmt::Display for EnvCheckError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::CheckFailed(e) => {
+                f.write_fmt(format_args!(
+                    "Type check failed with following error: {}",
+                    e
+                ))
+            }
+            Self::RegistryError(e) => {
+                f.write_fmt(format_args!(
+                    "Error occurred while parsing type registry: {}",
+                    e
+                ))
+            }
+        }
+    }
+}
+
+impl std::error::Error for EnvCheckError {}
+
+impl From<anyhow::Error> for EnvCheckError {
+    fn from(value: anyhow::Error) -> Self {
+        Self::RegistryError(value.to_string())
+    }
+}
```

Comment:
Could get rid of all of this if we just used `anyhow::Error` at the top level. You are already using that anyway on the internal methods.

Reply:
I want to have a concrete error for the failed field check

Reply:
Why though? It just gets converted immediately back into an `anyhow::Error` in the only (non test) usage: https://github.com/paritytech/cargo-contract/blob/9ebfa62c7d0ddc6e252461cddee21f00fe7d5845/crates/extrinsics/src/lib.rs#L383.

And for all the other errors you will have lost all the magic from the context and error chaining by converting the `anyhow::Error` to a `String` in `Self::RegistryError(value.to_string())`.